### ISSUE TITLE
Defer load the entire Flutter engine and framework

### ISSUE THREAD

**Issue by xster** [flutter/flutter#81265]
With #57617, we're now using Android's Play Feature Delivery mechanism (https://developer.android.com/guide/playcore/feature-delivery) to slice up the Dart code into separately downloadable "deferred components". 

Another way we can use the Android Play Feature Delivery mechanism is to defer load the entirety of Flutter rather than parts of the user's Flutter application. Specifically, the high initial fixed cost components like the Flutter engine, the Dart runtime, the Flutter framework etc can all be remote loaded at the time of use. 

Broadly, we can accomplish this by producing a simple separate maven artifact that only has Java classes to load the main Flutter "feature" via the Play Store APIs. Then we need tooling support and documentation to help users configure their Android and gradle project to package the main Flutter into a separate feature directory. For instance, the main Flutter .AAR can be built separately via `flutter build aar` than the remote loader helper class.

**Comment by kdcyberdude** [comment]
I am facing the same issue/crash after adding flutter AAR in the on-demand dynamic module [#56284](https://github.com/flutter/flutter/issues/56284)[Same crash logs]. Our team is working to launch premium functionality on the existing native application as a Flutter “On-demand module”. Due to the existing APK size, Million monthly active users and given requirements, additional overhead size increase due to the Flutter is a big concern. So I have tried adding it as an on-demand module but the app is crashing as soon as we create an instance of FlutterEngine(context).   I am trying to understand what causes it and how can I solve this. So far I am able to deduce…

**Flutter as an on-demand module is working in the below cases -**
- Debug variant.
- Release build is working only after adding **libapp.so** files in app module. Specifically in 
app/source/main/jniLibs/{cpu_architecture}/libapp.so [ Which is not suitable in our case because it’s redundant ]
- Changing the configuration of a dynamic feature module from “on-demand” to “install-time”.

The crash is specifically happening when FlutterEngine constructor calls **nativeAttach()** function present in FlutterJNI.java
Function signature is - line 319 | private native long nativeAttach(FlutterJNI flutterJNI, boolean isBackgroundView);


As mentioned here - https://github.com/android/app-bundle-samples/issues/30#issuecomment-644482601
> > > i'm using Dynamic Feature Module for a feature which has third SDK library as `.AAR` file.
> > > When analysing the apk i found out this sdk is using Native / C++ codes and the structure is like:
> > > ![Screen Shot 2019-09-18 at 2 43 20 pm](https://user-images.githubusercontent.com/31118131/65402931-df628a00-de14-11e9-99f5-708c299a9eb3.png)
> > > i know that in dynamic module when we want to use native libs we should implement for instance something like :`SplitInstallHelper.loadLibrary(this, "tensorflowlite_gpu_gl")`
> > > however i'm getting this error:
> > > java.lang.UnsatisfiedLinkError: dlopen failed: library "libtensorflowlite_gpu_gl.so" not found
> > > I'm not sure in Gradle in dynamic module what should I do to be able to load those .so libs?
> > 
> > 

> 
> It's better to change the `.ARR` file rather than the app, also try to use `SplitCompat.installActivity(this)` instead of just calling `SplitCompat.install(this)` it helps immediate access to a given activity.


Flutter AAR might needs to be updated to call SplitInstallHelper.loadLibrary(applicationContext, “app”) as mentioned above.

After trying this on the “on-demand module”

> SplitInstallHelper.loadLibrary(applicationContext, “app”) and SplitInstallHelper.loadLibrary(applicationContext, “flutter”) 
It seems it is able to load successfully because I did not get **UnsatisfiedLinkError** but the app still crashes.

After that, I tried loading the library from the path

> System.load("/data/data/com.package.name/files/splitcompat/3/native-libraries/flutter.config.arm64_v8a”) 
It seems to be able to load successfully as well and crash does happen as well. I have called this before creating an instance of **FlutterEngine()**

I would be really happy if anyone is able to help me in any way. Thanks in advance.
You guys might be able to help. I have seen you guys working on the internals of Flutter engine @GaryQian @xster @chinmaygarde @jason-simmons

**Comment by ldasilva-net** [comment]
It's seems related to this issue too  https://github.com/flutter/flutter/issues/92061

**Comment by xster** [comment]
We can close this now right?

**Comment by kdcyberdude** [comment]
I am able to understand the core of this problem. And able to do some workaround to make it work. 

The problem is in the flutter android embedding layer specifically in the FlutterLoader.java file. It has one private variable named as flutterApplicationInfo which initializes in `startInitialization()` function with an object of FlutterApplicationInfo which get's it's values form `getApplicationInfo(applicationContext)` which has `nativeLibraryDir` and points to base module libs directory. 

That's why it was working (mentioned in the previous comment)

> Release build is working only after adding **libapp.so** files in app module. Specifically in
> app/source/main/jniLibs/{cpu_architecture}/**libapp.so** [ Which is not suitable in our case because it’s redundant ]

Instead of patching the engine locally, I used Reflection to solve this problem. With the help of reflection, I did update the flutterApplicationInfo private variable of FlutterLoader.java as shown below,

  ```
fun initializeFlutterEngine(applicationContext: Context) {
        if (initialized) {
            Log.d(TAG, "Flutter initialize Engine is already called")
            return
        }

        Log.d(TAG, "Initializing Flutter Engine")
        
        var flutterLibsPath = ""

        val searchFiles: Queue<File> = LinkedList()
        // Downloaded modules are stored here
        searchFiles.add(applicationContext.filesDir)

        val aotSharedLibraryName = "libapp.so"

        while (!searchFiles.isEmpty()) {
            val file = searchFiles.remove()
            if (file != null && file.isDirectory && file.listFiles() != null) {
                for (f in file.listFiles()) {
                    searchFiles.add(f)
                }
                continue
            }
            val name = file!!.name

            if (name == aotSharedLibraryName) {
                val path = file.absolutePath.dropLast(aotSharedLibraryName.length)
                Log.d(TAG, "Flutter Library Path: $path")
                flutterLibsPath = path
                flutterLibsPathExist = true
            }
        }

        if (!flutterLibsPathExist) {
            Log.d(
                TAG, "Unable to get library paths of Flutter!. Maybe this is a fat APK or Flutter module is loaded as install time module"
            )
        }

        val injector = FlutterInjector.instance()
        val flutterLoader = injector.flutterLoader()

        flutterLoader.startInitialization(applicationContext)

        val flutterApplicationInfo = flutterLoader.javaClass.getDeclaredField("flutterApplicationInfo")
        flutterApplicationInfo.isAccessible = true

        val appInfo = getApplicationInfo(applicationContext) as ApplicationInfo
       // getString(), getNetworkPolicy(), GetBoolean() is copied from FlutterLoader.java 
        val res = FlutterApplicationInfo(
            getString(appInfo.metaData, ApplicationInfoLoader.PUBLIC_AOT_SHARED_LIBRARY_NAME),
            getString(appInfo.metaData, ApplicationInfoLoader.PUBLIC_VM_SNAPSHOT_DATA_KEY),
            getString(appInfo.metaData, ApplicationInfoLoader.PUBLIC_ISOLATE_SNAPSHOT_DATA_KEY),
            getString(appInfo.metaData, ApplicationInfoLoader.PUBLIC_FLUTTER_ASSETS_DIR_KEY),
            getNetworkPolicy(appInfo, applicationContext),
            flutterLibsPath.ifEmpty { appInfo.nativeLibraryDir },
            getBoolean(appInfo.metaData, PUBLIC_AUTOMATICALLY_REGISTER_PLUGINS_METADATA_KEY, true)
        )


        flutterApplicationInfo.set(flutterLoader, res)

        flutterLoader.ensureInitializationComplete(applicationContext, null)

        val jni = injector.flutterJNIFactory.provideFlutterJNI()

        flutterEngine = FlutterEngine(applicationContext, flutterLoader, jni)

        // Start executing Dart code to pre-warm the FlutterEngine.
        flutterEngine.dartExecutor.executeDartEntrypoint(
            DartExecutor.DartEntrypoint.createDefault()
        )

        // Cache the FlutterEngine to be used by FlutterActivity.
        FlutterEngineCache
            .getInstance()
            .put(AppConstants.FLUTTER_MAIN_ENGINE_ID, flutterEngine)

        initialized = true
    }

```

After doing this the app will not crash. Although there are going to be some other problems - 
**We cannot access any assets!!** 
I have tried to update the asset manager with `flutterJNI.updateJavaAssetManager(assetManager, flutterApplicationInfo.flutterAssetsDir);` as provided in flutter 2.8.1 because of a **deferred component** feature this will lead to a crash with log-

> backtrace:
> 2022-03-11 17:04:02.728 23592-23592/? A/DEBUG:     NOTE: Function names and BuildId information is missing for some frames due
> 2022-03-11 17:04:02.728 23592-23592/? A/DEBUG:     NOTE: to unreadable libraries. For unwinds of apps, only shared libraries
> 2022-03-11 17:04:02.728 23592-23592/? A/DEBUG:     NOTE: found under the lib/ directory are readable.
> 2022-03-11 17:04:02.728 23592-23592/? A/DEBUG:       #00 pc 000000000053e1f4  /data/data/com.tickledmedia.ParentTown/files/splitcompat/650/native-libraries/flutter.config.arm64_v8a/libflutter.so
> 2022-03-11 17:04:02.728 23592-23592/? A/DEBUG:       01 pc 000000000004c0ac  [anon:libc_malloc]
> * */

This might require additional investigation and a flutter engine patch to make it work. I did not explore this further because it is creating problems in deferred component feature as well, Github issues associated with that-
[Unable to load deferred component assets #96046](https://github.com/flutter/flutter/issues/96046)
[Lost image assets from downloaded deferred component on release build #86923](https://github.com/flutter/flutter/issues/86923)

To solve this I have loaded all assets from the network as a zip file and unarchived that to use in the application.

Now your application can load all the assets!!
**The other problem that is going to occur is you are not able to see any material design icons. That you or libraries have used in your application using Icons.whatever.** 
The workaround for this is to create an assets folder in your app module and add the files from the flutter module after extracting them from the app bundle as shown in the image.
<img width="251" alt="image" src="https://user-images.githubusercontent.com/34835322/159657099-baa2abe6-33f4-41b2-b74d-aa1f875facb9.png">

**Comment by ldasilva-net** [comment]
Hi @xster. 
I think that @kdcyberdude hit the nail on the head with his discovery. You can find an implementation example here: https://github.com/flutter/flutter/issues/92061

Can the Flutter team analyze it? 
Thanks a lot!

**Comment by liulianshanzhu** [comment]
```
val injector = FlutterInjector.instance()
val flutterLoader = injector.flutterLoader()
flutterLoader.startInitialization(applicationContext)
val pathLoader = applicationContext.classLoader as PathClassLoader
val path = pathLoader.findLibrary("app")
val nativePtah = pathLoader.findLibrary("app").dropLast("/libapp.so".length)
val shell = "--aot-shared-library-name=$path"
flutterLoader.ensureInitializationComplete(applicationContext, arrayOf(shell))
```
it's ok.

**Comment by kdcyberdude** [comment]
@liulianshanzhu It works like a charm. Awesome dude. You should write a blog post about "dart VM flags".😎
@ldasilva-net Created a PR for other to test as well. https://github.com/ldasilva-net/feature_delivery_flutter/pull/2

cc: @GaryQian

**Comment by reidbaker** [comment]
Gary is no longer working on flutter issues. Removing him to correctly reflect bug status.
